lib/jxl/dec_external_image.cc

```diff
@@ -67,30 +70,29 @@ void FloatToU32(const float* in, uint32_t* out, size_t num, float mul,
       out[x] = 0;
     }
   }
+
+  // Poison back the output.
+  PoisonMemory(out + num, sizeof(out[0]) * (num_round_up - num));
 }
 
 void FloatToF16(const float* in, hwy::float16_t* out, size_t num) {
   const HWY_FULL(float) d;
   const hwy::HWY_NAMESPACE::Rebind<hwy::float16_t, decltype(d)> du;
-  size_t vec_num = num;
-#if JXL_IS_DEBUG_BUILD
-  // Avoid accessing partially-uninitialized vectors with memory sanitizer.
-  vec_num &= ~(Lanes(d) - 1);
-#endif  // JXL_IS_DEBUG_BUILD
-  for (size_t x = 0; x < vec_num; x += Lanes(d)) {
+
+  // Unpoison accessing partially-uninitialized vectors with memory sanitizer.
+  // This is because we run DemoteTo() on the vector which triggers msan on
+  // scalar implementation.
```

Comment:
Uhm, this feels wrong - unpoison/poison should be a noop on scalar, no?

Reply:
Hopefully they are. Those are no-op in non-MSAN build; and semantical no-op for scalar MSAN build (num_round_up == num)

Reply:
Right, but then the comment is suspicious...

Reply:
Yup, it will be more clear if we mention that transformation is lane-wise and DemoteTo() might use comparison results in computation (and MSAN hates that).

Reply:
hmmm.. well the DemoteTo implementation in scalar definitively triggers msan because of its if(); it is possible that the simd implementation of DemoteTo doesn't; but it is true that in scalar this would be a no-op. We were already processing the last vector separately in msan so I would keep this code here but will update the comment.